### ISSUE TITLE
Problem in webserver with boost getrandom

### ISSUE THREAD

**Issue by Vaufreyd** [domoticz/domoticz#2982]
Hello,

I am using domoticz since 2 years on a chromebox with Linux. It was working fine and get no trouble. I want to upgrade to current version (compiling from source went fine) and get a strange behavior when connecting to the web page:
2019-01-19 10:59:41.916  Error: [web:8080] exception occurred : 'getrandom' (need to run again)
2019-01-19 10:59:41.916  Error: WebServer(HTTP) exception occurred : 'getrandom' 

It seems to come from the block starting at line 1335 of  webserver/cWebem.cpp file (when generating randomly a session key). 

I tried to run it as root, same problem.

Does anyone encounter this problem?

Best regards.

**Comment by gizmocuz** [comment]
Replaced the boost code, can you test if this solved the problem ?

**Comment by kofec** [comment]
I also have this issue. With this commit it change to:

```
[ 73%] Building CXX object CMakeFiles/domoticz.dir/hardware/Thermosmart.cpp.o
[ 74%] Building CXX object CMakeFiles/domoticz.dir/hardware/ToonThermostat.cpp.o
[ 74%] Building CXX object CMakeFiles/domoticz.dir/hardware/TTNMQTT.cpp.o
......
[100%] Linking CXX executable domoticz
/mnt/openwrt/OpenwrtAR71xxUSB/tmp/cc6sKQ5G.ltrans13.ltrans.o: In function `CToonThermostat::GetRandom[abi:cxx11]()':
<artificial>:(.text+0x5ebc): undefined reference to `getrandom'
collect2: error: ld returned 1 exit status
CMakeFiles/domoticz.dir/build.make:3621: recipe for target 'domoticz' failed
make[5]: *** [domoticz] Error 1
make[5]: Leaving directory '/mnt/openwrt/OpenwrtAR71xxUSB/build_dir/target-mips_24kc_musl/domoticz-454811f7196b77b260abafb7c6dbfda2437fae48'
CMakeFiles/Makefile2:70: recipe for target 'CMakeFiles/domoticz.dir/all' failed
make[4]: *** [CMakeFiles/domoticz.dir/all] Error 2
make[4]: Leaving directory '/mnt/openwrt/OpenwrtAR71xxUSB/build_dir/target-mips_24kc_musl/domoticz-454811f7196b77b260abafb7c6dbfda2437fae48'
Makefile:151: recipe for target 'all' failed
make[3]: *** [all] Error 2
make[3]: Leaving directory '/mnt/openwrt/OpenwrtAR71xxUSB/build_dir/target-mips_24kc_musl/domoticz-454811f7196b77b260abafb7c6dbfda2437fae48'
Makefile:137: recipe for target '/mnt/openwrt/OpenwrtAR71xxUSB/build_dir/target-mips_24kc_musl/domoticz-454811f7196b77b260abafb7c6dbfda2437fae48/.built' failed
make[2]: *** [/mnt/openwrt/OpenwrtAR71xxUSB/build_dir/target-mips_24kc_musl/domoticz-454811f7196b77b260abafb7c6dbfda2437fae48/.built] Error 2
make[2]: Leaving directory '/mnt/openwrt/OpenwrtAR71xxUSB/feeds/packages/utils/domoticz'
time: package/feeds/packages/domoticz/compile#1219.92#86.72#1274.67
package/Makefile:107: recipe for target 'package/feeds/packages/domoticz/compile' failed
make[1]: *** [package/feeds/packages/domoticz/compile] Error 2
make[1]: Leaving directory '/mnt/openwrt/OpenwrtAR71xxUSB'
/mnt/openwrt/OpenwrtAR71xxUSB/include/toplevel.mk:216: recipe for target 'package/domoticz/compile' failed
make: *** [package/domoticz/compile] Error 2
```
it is branch (18.06) of Openwrt for toolchain-mips_24kc_gcc-7.3.0_musl.

When build for master/trunk OpenWrt it is fine (there is only warring):

```
[ 73%] Building CXX object CMakeFiles/domoticz.dir/hardware/Thermosmart.cpp.o
[ 74%] Building CXX object CMakeFiles/domoticz.dir/hardware/ToonThermostat.cpp.o
In file included from /mnt/openwrt/OpenwrtXiaomi/staging_dir/target-mipsel_24kc_musl/usr/include/boost/random/detail/integer_log2.hpp:19:0,
                 from /mnt/openwrt/OpenwrtXiaomi/staging_dir/target-mipsel_24kc_musl/usr/include/boost/random/detail/large_arithmetic.hpp:19,
                 from /mnt/openwrt/OpenwrtXiaomi/staging_dir/target-mipsel_24kc_musl/usr/include/boost/random/detail/const_mod.hpp:23,
                 from /mnt/openwrt/OpenwrtXiaomi/staging_dir/target-mipsel_24kc_musl/usr/include/boost/random/detail/seed_impl.hpp:26,
                 from /mnt/openwrt/OpenwrtXiaomi/staging_dir/target-mipsel_24kc_musl/usr/include/boost/random/mersenne_twister.hpp:30,
                 from /mnt/openwrt/OpenwrtXiaomi/staging_dir/target-mipsel_24kc_musl/usr/include/boost/uuid/random_generator.hpp:17,
                 from /mnt/openwrt/OpenwrtXiaomi/staging_dir/target-mipsel_24kc_musl/usr/include/boost/uuid/uuid_generators.hpp:17,
                 from /mnt/openwrt/OpenwrtXiaomi/build_dir/target-mipsel_24kc_musl/domoticz-454811f7196b77b260abafb7c6dbfda2437fae48/hardware/ToonThermostat.cpp:14:
/mnt/openwrt/OpenwrtXiaomi/staging_dir/target-mipsel_24kc_musl/usr/include/boost/pending/integer_log2.hpp:7:59: note: #pragma message: This header is deprecated. Use <boost/integer/integer_log2.hpp> instead.
 BOOST_HEADER_DEPRECATED("<boost/integer/integer_log2.hpp>");
                                                           ^
[ 74%] Building CXX object CMakeFiles/domoticz.dir/hardware/TTNMQTT.cpp.o
```

**Comment by Vaufreyd** [comment]
Hello,

I also change the code not to use boost at this point. It works (but I fall into another problem of dead lock after a while). I will try to use your proposal (integrating only your commit) to see is it works for me and keep you informed (maybe in a few days).

Best regards. Doms.

PS: this also questionned the need for boost as it is a strong dependancy (your version is not the stable version on my Linux, it was tricky to compile it on my system). With c++11, as far as I saw, it could be replaced easily (for random, for thread, I need to check if domoticz is using other features from boost).

**Comment by gizmocuz** [comment]
@Vaufreyd , yes i am constantly trying to remove non-needed boost code... 
If you would like to help, please go ahead ;)

Anyways, we use this boost version for a long time now! Never received any feedback that it caused issues, so i find it hard to believe it is all depended on this boost version

Are you following the build guide on our wiki to compile for your linux system ?

https://www.domoticz.com/wiki/Raspberry_Pi_-_Build_Domoticz_from_source

**Comment by Vaufreyd** [comment]
@gizmocuz, If I manage to get time to work on it, I will.

About boost, it has no intrinsec issue but is a dependancy that can be advantageously removed with new c++ features. In my case (debian like Linux for Chromebox), I can not install the needed version using apt. I followed the tutorial, but need to patch the cmake (and some system files) to make it compile...  

Next step, try your code for the key session generation.

**Comment by gizmocuz** [comment]
in the wiki it is shown how to build the 3 (i guess it is 3) boost libraries that are needed, no need to apt, and no need to patch the cmake script
Depending on your architecture, you could also try one of the binaries that our build servers generate (arm 32/64, linux 64, osx)

Yes i know these days a lot of boost code can be rewritten with the latest cxx libraries, i try to avoid as much boost code as possible, but sometimes boost is still superior ....

**Comment by Vaufreyd** [comment]
First, about the test of your code. No more problem for me with the boost random generator.

Second, about compilation. I am computer scientist, thus I understand the process of compilation. On my debian like system, the installation of the boost library (actually, includes and few compile code as .so and .a) is not functionning (Compilation went ok, installatin seems ok but compiling against this version of boost is not working and I do not find why). I tried : 
- follow strickly the wiki (as I did several times before) => failed (with boost problem)
- try to install 1.68.0 with apt => failed (not available on my ditrib for the moment, I have 1.58.0)
- change a little bit the compilation schema (see below)=> ok.

This is what I did (it may help someone, maybe it could be added to the wiki as a workaround...) : 
- first install boost in a local folder called 'boost_install' instead of system (as it failed as I said)
./b2 install threading=multi link=static --with-thread --with-system --prefix=../boost_install/
- compile saying CMAKE where to find boost with option
cmake -DCMAKE_BUILD_TYPE=Release -DBoost_NO_BOOST_CMAKE=TRUE -DBOOST_ROOT=../boost_install/ -DBOOST_LIBRARYDIR=../boost_install/lib/ .
=> it failed (did not find the libboost_system.a file...) even if the -DBOOST_LIBRARYDIR should have done the trick
- add this line to the CMakeList.txt
link_directories("/home/user/newdomo/boot_install/lib")
- relaunch the previous cmake command and it works

Doms.
